scapy/contrib/gtp_v2.py

```diff
@@ -0,0 +1,979 @@
+            l = len(p)-8
+            p = p[:2] + struct.pack("!H", l) + p[4:]
+        return p
+
+    def hashret(self):
+        return struct.pack("B", self.version) + self.payload.hashret()
+
+    def answers(self, other):
+        return (isinstance(other, GTPHeader) and
+                self.version == other.version and
+                self.payload.answers(other.payload))
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt and len(_pkt) >= 1:
+            if (struct.unpack("B", _pkt[0])[0] >> 5) & 0x7 == 1:
+                return gtp.GTPHeader
+        return GTPHeader
+
+
+class IE_IPv4(gtp.IE_Base):
+    name = "IE IPv4"
+    fields_desc = [ByteEnumField("ietype", 154, IEType),
+                   ShortField("length", 0),
+                   BitField("CR_flag", 0, 4),
+                   BitField("instance", 0, 4),
+                   IPField("address", RandIP())]
+
+
+class IE_MEI(gtp.IE_Base):
+    name = "IE MEI"
+    fields_desc = [ByteEnumField("ietype", 154, IEType),
+                   ShortField("length", 0),
+                   BitField("CR_flag", 0, 4),
+                   BitField("instance", 0, 4),
+                   LongField("MEI", 0)]
+
+
+def IE_Dispatcher(s):
+    """Choose the correct Information Element class."""
+    if not s:
+        return Raw(s)
+
+    # Get the IE type
+    ietype = ord(s[0])
+    cls = ietypecls.get(ietype, Raw)
+
+    # if ietype greater than 128 are TLVs
+    if cls is Raw and ietype > 128:
+        cls = IE_NotImplementedTLV
+
+    return cls(s)
+
+
+class IE_EPSBearerID(gtp.IE_Base):
+    name = "IE EPS Bearer ID"
+    fields_desc = [ByteEnumField("ietype", 154, IEType),
+                   ShortField("length", 0),
+                   BitField("CR_flag", 0, 4),
+                   BitField("instance", 0, 4),
+                   ByteField("EBI", 0)]
+
+
+class IE_RAT(gtp.IE_Base):
+    name = "IE RAT"
+    fields_desc = [ByteEnumField("ietype", 154, IEType),
+                   ShortField("length", 0),
+                   BitField("CR_flag", 0, 4),
+                   BitField("instance", 0, 4),
+                   ByteEnumField("RAT_type", None, RATType)]
+
+
+class IE_ServingNetwork(gtp.IE_Base):
+    name = "IE Serving Network"
+    fields_desc = [ByteEnumField("ietype", 154, IEType),
+                   ShortField("length", 0),
+                   BitField("CR_flag", 0, 4),
+                   BitField("instance", 0, 4),
+                   gtp.TBCDByteField("MCC", "", 2),
+                   gtp.TBCDByteField("MNC", "", 1)]
+
+
+class ULI_RAI(gtp.IE_Base):
+    name = "IE Tracking Area Identity"
+    fields_desc = [  # Only type 1 is currently supported
+        gtp.TBCDByteField("MCC", "", 2),
+        # MNC: if the third digit of MCC is 0xf, then the length of
+        # MNC is 1 byte
+        gtp.TBCDByteField("MNC", "", 1),
+        ShortField("LAC", 0),
+        ShortField("SAC", 0)]
+
+
+class ULI_SAI(gtp.IE_Base):
```

Comment:
Is it normal that ULI_SAI and ULI_RAI are exactly the same ? Can you combine them ?